manticore/core/smtlib/solver.py

```diff
@@ -588,62 +620,107 @@ def _optimize_fancy(self, constraints: ConstraintSet, x: BitVec, goal: str, max_
                         var.append(subvar)
                         temp_cs.add(subvar == simplify(expression[i]))
-                    self._reset(temp_cs.to_string())
-                    if not self._is_sat():
-                        raise SolverError(
-                            "Solver could not find a value for expression under current constraint set"
-                        )
-
-                    for i in range(expression.index_max):
-                        result.append(self.__getvalue_bv(var[i].name))
-                    values.append(bytes(result))
-                    if time.time() - start > consts.timeout:
-                        raise SolverError("Timeout")
-                    continue
+                    vars.append(var)
 
-                temp_cs.add(var == expression)
-
-                self._reset(temp_cs.to_string())
+            self._reset(temp_cs.to_string())
 
-                if not self._is_sat():
-                    raise SolverError(
-                        "Solver could not find a value for expression under current constraint set"
-                    )
+            if not self._is_sat():
+                raise SolverError(
+                    "Solver could not find a value for expression under current constraint set"
+                )
 
+            values_to_ask: List[str] = []
+            is_bv: List[bool] = []
+            for idx, expression in enumerate(expressions):
+                if not issymbolic(expression):
+                    continue
+                var = vars[idx]
                 if isinstance(expression, Bool):
-                    values.append(self.__getvalue_bool(var.name))
+                    values_to_ask.append(var.name)
+                    is_bv.append(False)
                 if isinstance(expression, BitVec):
-                    values.append(self.__getvalue_bv(var.name))
+                    values_to_ask.append(var.name)
+                    is_bv.append(True)
+                if isinstance(expression, Array):
+                    # result = []
+                    for i in range(expression.index_max):
+                        # result.append(self.__getvalue_bv(var[i].name))
+                        values_to_ask.append(var[i].name)
+                        is_bv.append(True)
+
+            if values_to_ask == []:
+                return values
+
+            values_returned, sol = self.__getvalue_all(values_to_ask, is_bv)
+            # print(values_returned)
+            # print(values_returned.keys())
+            # print(values_returned)
+            for idx, expression in enumerate(expressions):
+                if not issymbolic(expression):
+                    continue
+                var = vars[idx]
+                if isinstance(expression, Bool):
+                    values[idx] = values_returned[var.name]
+                if isinstance(expression, BitVec):
+                    if var.name not in values_returned:
+                        print("Error")
```

Comment:
We should probably import the logging module and use `logger.error` here

Reply:
Done.